core/domain/prod_validation_jobs_one_off.py

```diff
@@ -110,67 +351,3183 @@ def validate(cls, item):
+                    }
+                }]
+            }
+        }
+
+        return {
+            'activity_references': activity_references_schema
+        }
+
+    @classmethod
+    def _get_model_domain_object_instances(cls, item):
+        model_domain_object_instances = []
+
+        try:
+            for reference in item.activity_references:
+                model_domain_object_instances.append(
+                    activity_domain.ActivityReference(
+                        reference['type'], reference['id']))
+        except Exception as e:
+            cls.errors['fetch properties'].append((
+                'Entity id %s: Entity properties cannot be fetched completely '
+                'with the error %s') % (item.id, e))
+            return []
+
+        return model_domain_object_instances
+
+    @classmethod
+    def _get_commit_cmd_domain_class(cls):
+        return None
+
+    @classmethod
+    def _get_external_id_relationships(cls, item):
+        exploration_ids = []
+        collection_ids = []
+
+        try:
+            for reference in item.activity_references:
+                if reference['type'] == constants.ACTIVITY_TYPE_EXPLORATION:
+                    exploration_ids.append(reference['id'])
+                elif reference['type'] == constants.ACTIVITY_TYPE_COLLECTION:
+                    collection_ids.append(reference['id'])
+        except Exception as e:
+            cls.errors['fetch properties'].append((
+                'Entity id %s: Entity properties cannot be fetched completely '
+                'with the error %s') % (item.id, e))
+            return {}
+
+        return {
+            'exploration_ids': (exp_models.ExplorationModel, exploration_ids),
+            'collection_ids': (
+                collection_models.CollectionModel, collection_ids),
+        }
+
+    @classmethod
+    def _get_custom_validation_functions(cls):
+        return []
+
+
+class RoleQueryAuditModelValidator(BaseModelValidator):
+    """Class for validating RoleQueryAuditModels."""
+
+    @classmethod
+    def _get_model_id_regex(cls, item):
+        # Valid id: [user_id].[timestamp_in_sec].[intent].[random_number]
+        regex_string = '%s\\.\\d*\\.%s\\.\\d*$' % (item.user_id, item.intent)
+        return regex_string
+
+    @classmethod
+    def _get_json_properties_schema(cls, item):
+        return {}
+
+    @classmethod
+    def _get_model_domain_object_instances(cls, item):
+        return []
+
+    @classmethod
+    def _get_commit_cmd_domain_class(cls):
+        return None
+
+    @classmethod
+    def _get_external_id_relationships(cls, item):
+        return {'user_ids': (user_models.UserSettingsModel, [item.user_id])}
+
+    @classmethod
+    def _get_custom_validation_functions(cls):
+        return []
+
+
+class CollectionModelValidator(BaseModelValidator):
+    """Class for validating CollectionModel."""
+
+    @classmethod
+    def _get_model_id_regex(cls, item):
+        return '.'
```

Comment:
Hi @seanlip, since there is no ^ and $ for the for the regex it will just match a character at the beginning and will pass for all the model ids.

I did this because I did not find any constraints for model ids as you mention that they are base64 strings. I am still not able to find any documentation for model ids being base64 strings. Do we have that documented somewhere, if yes can you please direct me to that documentation. Thanks!

Reply:
You need to dig a bit, but look at the create() methods and in particular get_new_id() in core/storage/base_model.

Reply:
Thanks for this. I checked that get_new_id produces ids as base64 strings with max length as 12. Updated the code now.